### ISSUE TITLE
Crash in -[SentryCrashReportConverter threadAtIndex:stripCrashedStacktrace:]

### ISSUE THREAD

**Issue by lhunath** [getsentry/sentry-cocoa#364]
_Platform:_
- [x]  iOS
- [ ]  tvOS
- [ ]  MacOS
- [ ]  watchOS

_Swift:_
- [x]  Yes -> If yes, which version?
- [ ]  No

_sentry-cocoa installed with:_
- [x] CocoaPods
- [ ] Carthage
- [ ] Manually

Version of sentry-cocoa:
**4.4.2**

---
I have following issue:

Relaunching after an app crash forced with `fatalError()` in Swift, Sentry crashes the app on start-up.

**Steps to reproduce:**
1. Crash the app using `fatalError`
2. Relaunch the app.

**Actual result:**
Sentry crashes on start-up:
```
2020-01-07 12:33:37.463938-0500 Volto[8368:3229909] -[NSTaggedPointerString objectForKeyedSubscript:]: unrecognized selector sent to instance 0xd1f2f75b7dedffa6
```

**Expected result:**
Sentry posts the `fatalError` crash report to sentry.io and allows the app to resume.

**Details:**
It's possible the issue is related to also having Bugsnag initialized.

Method at cause:
```
- (SentryThread *_Nullable)threadAtIndex:(NSInteger)threadIndex stripCrashedStacktrace:(BOOL)stripCrashedStacktrace {
    if (threadIndex >= [self.threads count]) {
        return nil;
    }
    NSDictionary *threadDictionary = [self.threads objectAtIndex:threadIndex];

    SentryThread *thread = [[SentryThread alloc] initWithThreadId:threadDictionary[@"index"]]; // <<< CRASH POINT
    // We only want to add the stacktrace if this thread hasn't crashed
    thread.stacktrace = [self stackTraceForThreadIndex:threadIndex];
    if (stripCrashedStacktrace && [threadDictionary[@"crashed"] boolValue]) {
        thread.stacktrace = nil;
    }
    thread.crashed = threadDictionary[@"crashed"];
    thread.current = threadDictionary[@"current_thread"];
    thread.name = threadDictionary[@"name"];
    if (nil == thread.name) {
        thread.name = threadDictionary[@"dispatch_queue"];
    }
    return thread;
}
```

At the crash point, `threadDictionary` holds an `NSString`, value `@"index"`, not an `NSDictionary`.

Stack:
```
Thread 4 Queue : com.apple.root.default-qos (concurrent)
#0	0x00000001a1bbd06c in objc_exception_throw ()
#1	0x00000001a1d9843c in -[NSObject(NSObject) doesNotRecognizeSelector:] ()
#2	0x00000001a1e98e08 in ___forwarding___ ()
#3	0x00000001a1e9abec in _CF_forwarding_prep_0 ()
#4	0x0000000102d0e284 in -[SentryCrashReportConverter threadAtIndex:stripCrashedStacktrace:] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/Sentry/SentryCrashReportConverter.m:222
#5	0x0000000102d17150 in -[SentryCrashReportConverter convertThreads] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/Sentry/SentryCrashReportConverter.m:421
#6	0x0000000102d07a70 in -[SentryCrashReportConverter convertReportToEvent] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/Sentry/SentryCrashReportConverter.m:92
#7	0x0000000102d2f250 in __52-[SentryCrashReportSink filterReports:onCompletion:]_block_invoke at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/Sentry/SentryCrashReportSink.m:60
#8	0x0000000101b39368 in __wrap_dispatch_async_block_invoke ()
#9	0x0000000103009828 in _dispatch_call_block_and_release ()
#10	0x000000010300ac04 in _dispatch_client_callout ()
#11	0x000000010300d300 in _dispatch_queue_override_invoke ()
#12	0x000000010301c168 in _dispatch_root_queue_drain ()
#13	0x000000010301ca90 in _dispatch_worker_thread2 ()
#14	0x00000001a1bb1f64 in _pthread_wqthread ()
Enqueued from com.apple.main-thread (Thread 1) Queue : com.apple.main-thread (serial)
#0	0x000000010300f058 in dispatch_async ()
#1	0x0000000101b39288 in wrap_dispatch_async ()
#2	0x0000000102d2eb1c in -[SentryCrashReportSink filterReports:onCompletion:] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/Sentry/SentryCrashReportSink.m:54
#3	0x0000000102d225f0 in __62-[SentryCrashReportFilterPipeline filterReports:onCompletion:]_block_invoke_3 at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/SentryCrash/Reporting/Filters/SentryCrashReportFilterBasic.m:336
#4	0x0000000102d18034 in sentrycrash_callCompletion at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/SentryCrash/Reporting/Filters/SentryCrashReportFilter.h:72
#5	0x0000000102d20730 in -[SentryCrashReportFilterPipeline filterReports:onCompletion:] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/SentryCrash/Reporting/Filters/SentryCrashReportFilterBasic.m:292
#6	0x0000000102d21644 in -[SentryCrashReportFilterPipeline filterReports:onCompletion:] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/SentryCrash/Reporting/Filters/SentryCrashReportFilterBasic.m:348
#7	0x0000000102c775b0 in -[SentryCrash sendReports:onCompletion:] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/SentryCrash/Recording/SentryCrash.m:447
#8	0x0000000102c74f18 in -[SentryCrash sendAllReportsWithCompletion:] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/SentryCrash/Recording/SentryCrash.m:348
#9	0x0000000102ca1adc in -[SentryCrashInstallation sendAllReportsWithCompletion:] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/SentryCrash/Installations/SentryCrashInstallation.m:355
#10	0x0000000102d4ea70 in -[SentryInstallation sendAllReportsWithCompletion:] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/Sentry/SentryInstallation.m:46
#11	0x0000000102d4e628 in -[SentryInstallation sendAllReports] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/Sentry/SentryInstallation.m:42
#12	0x0000000102c68afc in __43-[SentryClient startCrashHandlerWithError:]_block_invoke at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/Sentry/SentryClient.m:474
#13	0x000000010300ac04 in _dispatch_client_callout ()
#14	0x000000010300c4f4 in _dispatch_once_callout ()
#15	0x0000000102c688e4 in _dispatch_once [inlined] at /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS13.2.sdk/usr/include/dispatch/once.h:84
#16	0x0000000102c6880c in -[SentryClient startCrashHandlerWithError:] at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Pods/Sentry/Sources/Sentry/SentryClient.m:471
#17	0x000000010101b8c8 in MPTracker.().init() at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Source/iOS/Util/MPTracker.swift:25
#18	0x000000010101a878 in MPTracker.__allocating_init() ()
#19	0x000000010101a828 in globalinit_33_B089636E75B43B38890D854A880BF08C_func0 at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Source/iOS/Util/MPTracker.swift:13
#20	0x000000010300ac04 in _dispatch_client_callout ()
#21	0x000000010300c4f4 in _dispatch_once_callout ()
#22	0x00000001af277924 in swift_once ()
#23	0x000000010101a8c0 in MPTracker.shared.unsafeMutableAddressor at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Source/iOS/Util/MPTracker.swift:13
#24	0x0000000100aaae30 in MPAppDelegate.application(_:willFinishLaunchingWithOptions:) at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Source/iOS/MPAppDelegate.swift:20
#25	0x0000000100aac044 in @objc MPAppDelegate.application(_:willFinishLaunchingWithOptions:) ()
#26	0x00000001a5e98e2c in -[UIApplication _handleDelegateCallbacksWithOptions:isSuspended:restoreState:] ()
#27	0x00000001a5e9ac8c in -[UIApplication _callInitializationDelegatesWithActions:forCanvas:payload:fromOriginatingProcess:] ()
#28	0x00000001a5ea04c0 in -[UIApplication _runWithMainScene:transitionContext:completion:] ()
#29	0x00000001a567c810 in -[_UISceneLifecycleMultiplexer completeApplicationLaunchWithFBSScene:transitionContext:] ()
#30	0x00000001a5b0d3ec in _UIScenePerformActionsWithLifecycleActionMask ()
#31	0x00000001a567d2f8 in __101-[_UISceneLifecycleMultiplexer _evalTransitionToSettings:fromSettings:forceExit:withTransitionStore:]_block_invoke ()
#32	0x00000001a567cd50 in -[_UISceneLifecycleMultiplexer _performBlock:withApplicationOfDeactivationReasons:fromReasons:] ()
#33	0x00000001a567d124 in -[_UISceneLifecycleMultiplexer _evalTransitionToSettings:fromSettings:forceExit:withTransitionStore:] ()
#34	0x00000001a567c9e0 in -[_UISceneLifecycleMultiplexer uiScene:transitionedFromState:withTransitionContext:] ()
#35	0x00000001a5680efc in __186-[_UIWindowSceneFBSSceneTransitionContextDrivenLifecycleSettingsDiffAction _performActionsForUIScene:withUpdatedFBSScene:settingsDiff:fromSettings:transitionContext:lifecycleActionType:]_block_invoke_2 ()
#36	0x00000001a5a40028 in +[BSAnimationSettings(UIKit) tryAnimatingWithSettings:actions:completion:] ()
#37	0x00000001a5b265ac in _UISceneSettingsDiffActionPerformChangesWithTransitionContext ()
#38	0x00000001a5680c34 in __186-[_UIWindowSceneFBSSceneTransitionContextDrivenLifecycleSettingsDiffAction _performActionsForUIScene:withUpdatedFBSScene:settingsDiff:fromSettings:transitionContext:lifecycleActionType:]_block_invoke ()
#39	0x00000001a5b264a8 in _UISceneSettingsDiffActionPerformActionsWithDelayForTransitionContext ()
#40	0x00000001a5680aa0 in -[_UIWindowSceneFBSSceneTransitionContextDrivenLifecycleSettingsDiffAction _performActionsForUIScene:withUpdatedFBSScene:settingsDiff:fromSettings:transitionContext:lifecycleActionType:] ()
#41	0x00000001a54f5efc in __64-[UIScene scene:didUpdateWithDiff:transitionContext:completion:]_block_invoke ()
#42	0x00000001a54f4a64 in -[UIScene _emitSceneSettingsUpdateResponseForCompletion:afterSceneUpdateWork:] ()
#43	0x00000001a54f5c34 in -[UIScene scene:didUpdateWithDiff:transitionContext:completion:] ()
#44	0x00000001a5e9e840 in -[UIApplication workspace:didCreateScene:withTransitionContext:completion:] ()
#45	0x00000001a5a60a0c in -[UIApplicationSceneClientAgent scene:didInitializeWithEvent:completion:] ()
#46	0x00000001a6f2d994 in -[FBSSceneImpl _callOutQueue_agent_didCreateWithTransitionContext:completion:] ()
#47	0x00000001a6f52960 in __86-[FBSWorkspaceScenesClient sceneID:createWithParameters:transitionContext:completion:]_block_invoke.154 ()
#48	0x00000001a6f380f8 in -[FBSWorkspace _calloutQueue_executeCalloutFromSource:withBlock:] ()
#49	0x00000001a6f5261c in __86-[FBSWorkspaceScenesClient sceneID:createWithParameters:transitionContext:completion:]_block_invoke ()
#50	0x000000010300ac04 in _dispatch_client_callout ()
#51	0x000000010300e028 in _dispatch_block_invoke_direct ()
#52	0x00000001a6f77540 in __FBSSERIALQUEUE_IS_CALLING_OUT_TO_A_BLOCK__ ()
#53	0x00000001a6f7720c in -[FBSSerialQueue _queue_performNextIfPossible] ()
#54	0x00000001a6f77734 in -[FBSSerialQueue _performNextFromRunLoopSource] ()
#55	0x00000001a1e127e0 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ ()
#56	0x00000001a1e12738 in __CFRunLoopDoSource0 ()
#57	0x00000001a1e11ed0 in __CFRunLoopDoSources0 ()
#58	0x00000001a1e0d01c in __CFRunLoopRun ()
#59	0x00000001a1e0c8bc in CFRunLoopRunSpecific ()
#60	0x00000001abc78328 in GSEventRunModal ()
#61	0x00000001a5ea26d4 in UIApplicationMain ()
#62	0x0000000100ab08ec in main at /Users/lhunath/workspace/lyndir/MasterPassword/platform-darwin/Source/iOS/MPAppDelegate.swift:13
#63	0x00000001a1c97460 in start ()
```

**Comment by fetzig** [comment]
Hi @lhunath 

Sorry, but this triggers a lot of questions:

Can you reproduce this reliably? Does it still happen when you "deactivate" bugsnag? Can you send the complete stack trace (all threads), and if possible, the log? When/where/how are you initializing sentry-cocoa? Does your code interface with sentry-cocoa in any other way (despite `Client.shared?.startCrashHandler()`)?

Off topic, bonus question ;) Why are you using sentry _and_ bugsnag?

**Comment by lhunath** [comment]
- I can reproduce the issue.
- It only happens if the app crashes with Bugsnag installed.
- Sentry initialization code (triggered at the start of `-[MPAppDelegate application:willFinishLaunchingWithOptions:]`):
```
        if let sentryDSN = decrypt( secret: sentryDSN ) {
            do {
                let sentry = try Sentry.Client( dsn: sentryDSN )
                Sentry.Client.shared = sentry

                sentry.enabled = false
                sentry.enableAutomaticBreadcrumbTracking()
                sentry.extra = [ "device": self.deviceIdentifier, "owner": self.ownerIdentifier ]
                try sentry.startCrashHandler()
            }
            catch {
                err( "Couldn't install Sentry [>TRC]" )
                trc( "[>] %@", error )
            }
        }
```

Later on:
```
        if appConfig.diagnostics {
            Sentry.Client.shared?.enabled = true
        }
        else {
            Sentry.Client.shared?.enabled = false
        }
```

- Other Sentry interactions:
```
Sentry.Client.shared?.extra?["smartlook"] = smartlook
```
```
                    let event = Event( level: sentrySeverity )
                    event.message = record.message
                    event.logger = "mpw"
                    event.timestamp = record.occurrence
                    event.tags = [ "file": record.fileName, "line": "\(record.line)", "function": record.function ]
                    Sentry.Client.shared?.appendStacktrace( to: event )
                    Sentry.Client.shared?.send( event: event )
```
```
                    let breadcrumb = Breadcrumb( level: sentrySeverity, category: "mpw" )
                    breadcrumb.type = "log"
                    breadcrumb.message = record.message
                    breadcrumb.timestamp = record.occurrence
                    breadcrumb.data = [ "file": record.fileName, "line": "\(record.line)", "function": record.function ]
                    Sentry.Client.shared?.breadcrumbs.add( breadcrumb )
```
```
        let sentryBreadcrumb = Breadcrumb( level: .info, category: "event" )
        sentryBreadcrumb.type = "user"
        sentryBreadcrumb.message = name
        sentryBreadcrumb.data = eventParameters
        Sentry.Client.shared?.breadcrumbs.add( sentryBreadcrumb )
```
```
        Sentry.Client.shared?.user = Sentry.User( userId: userId )
        Sentry.Client.shared?.user?.username = userName
        Sentry.Client.shared?.user?.extra = userConfig
```
```
        Sentry.Client.shared?.user = nil
```

- Why sentry and bugsnag?  Mostly because I'm evaluating them to decide which one's API and features I like best.

**Comment by fetzig** [comment]
> * I can reproduce the issue.
> * It only happens if the app crashes with Bugsnag installed.

There you have it. A short look at the bugsnag-cocoa repo indicates that bugsnag uses KSCrash as well. Installing sentry and bugsnag at once is most likely the issue.

suggest you compare them differenlty. This might be of interest to you as well: https://www.crashprobe.com/

> * Sentry initialization code (triggered at the start of `-[MPAppDelegate application:willFinishLaunchingWithOptions:]`):

this part looks fine, but I don't know if you will run in other issues when initializing in `willFinishLaunchingWithOptions` instead of `didFinishLaunchingWithOptions`. Sentry docs (and all examples I know) use `didFinishLaunchingWithOptions`. Since KSCrash suggests that as well, I expect this to be the only proper and supported way to install/initialize the crash handler correctly.

> * Other Sentry interactions:

The code itself looks good. It depends where it's called.

> * Why sentry and bugsnag?  Mostly because I'm evaluating them to decide which one's API and features I like best.

As mentioned above, suggest you init/install only one, not both libs. Don't switch between them in runtime and install only one (only once). This is, in part, a requirement from KSCrash (which is used by sentry and bugsnag).

**Comment by fetzig** [comment]
I'm closing this issue. Feel free to reopen or create a new one if some new findings come up.

**Comment by lhunath** [comment]
I mean, ideally KSCrash / Sentry ought to play nice alongside other crash monitors or at the very least inform the developer of the fact that it's done something bad.

**Comment by fetzig** [comment]
Evidently this isn’t clear for everyone. Including activation of diagnostics a.s.o. (See your second issue). 

Will try and check if we can improve docs.sentry.io. reopen this as reminder.

**Comment by lhunath** [comment]
You could mention it in the docs, "don't use Sentry alongside any other KSCrash-based solutions, such as Bugsnag", but that feels very much like giving up and covering up the real problem with a, "hey, just don't bend that way".  Also, is the issue actually related to KSCrash or would it surface in other cases as well such as custom application code?

A real fix would ensure that this issue cannot resurface, ie. permitting graciously the use of concurrent crash handlers.

Also, it's really not the place of the docs to mention this but rather the place of the code to detect this issue and fail usefully as opposed to catastrophically.

**Comment by fetzig** [comment]
This has much to do with the way errors/exceptions/a.s.o. are handled in programming languages (in our case swift, c, c++, obj-c, obj-c++) and the cocoa ecosystem. Those have "limitations", some of them by design. And that results in certain "limitations" in the blocks that are build on top. I.e., KSCrash can't be installed twice in a single process/app. Can't imagine a use case for installing KSCrash (or bugsnag or sentry) twice, but that's just me. In my experience, it's clear to most users that you can't catch an exception twice (in most cases). Or you can't crash if the debugger is attached to the app (because it pauses the process to make sure you can check the internal state when the error happened).

I try to give you an example of why your proposal makes no sense to me. Hope this helps:

when you're offline or google.com is offline, or dns isn't working or your connections to slow for set timeout or any other issue that prevents resolving google.com into an IP address... `ping` response like this:
```
$ ping google.com 
ping: cannot resolve google.com: Unknown host
$
```
and it will most likely never respond like this:
```
$ ping google.com 
 * it looks like youre offline
 * your ethernet cable is not connected
 * dns is missconfigured
 * you have airplane mode activated
 * your wifi card is shut off
$
```

In our case, there are plenty of possible ways to prevent sentry from working in a useful fashion. 

IMHO, what you propose is out of scope (sentry should make checks/validations).

But maybe you can come up with a PR that makes me realize I'm wrong.

In any case, I will close this issue for now. Feel free to come back if new findings regarding this or something new comes up.

Thanks for caring and sharing.
